src/app/providers/PokemonProvider.tsx

```diff
@@ -0,0 +1,253 @@
+        if (isMounted && res) {
+          navigate(`?limit=10&offset=0`);
+          setApp((prevState) => ({
+            ...prevState,
+            defaultSearch: 'status' in res ? undefined : res,
+            isError: 'status' in res,
+            error: 'status' in res ? res : undefined,
+            isLoading: false,
+          }));
+        }
+      } catch (e) {
+        console.error('Caught on try - initializeData - ', e);
+        if (isMounted) {
+          setApp((prevState) => ({
+            ...prevState,
+            isLoading: false,
+            isError: true,
+            error: { status: 0, message: 'Unexpected error' },
+          }));
+        }
+      }
+    };
+    initializeData();
+
+    return () => {
+      isMounted = false;
+    };
+  }, [dataFromLs, navigate]);
+
+  const onChange = useCallback((e: ChangeEvent<HTMLInputElement>) => {
+    const { name, value } = e.target;
+    setApp((prevState) => ({
+      ...prevState,
+      [name]: value.toLowerCase().trim(),
+    }));
+  }, []);
+
+  const onSubmit = useCallback(
+    async (e: FormEvent<HTMLFormElement>) => {
+      e.preventDefault();
+      e.stopPropagation();
+      if (!search) {
+        setLocalState('');
+      }
+      try {
+        setLocalState(search);
+
+        if (search) {
+          setApp((prevState) => ({ ...prevState, isLoading: true }));
+          const res = await fetchDetailed(search);
+          setApp((prevState) => ({
+            ...prevState,
+            defaultSearch: undefined,
+            pokemon: 'status' in res ? undefined : res,
+            isError: 'status' in res,
+            error: 'status' in res ? res : undefined,
+            isLoading: false,
+          }));
+          return;
+        }
+        setApp((prevState) => ({ ...prevState, isLoading: true }));
+        const res = await fetchDefaultData();
+        setApp((prevState) => ({
+          ...prevState,
+          defaultSearch: 'status' in res ? undefined : res,
+          pokemon: undefined,
+          isError: 'status' in res,
+          error: 'status' in res ? res : undefined,
+          isLoading: false,
+        }));
+      } catch (e) {
+        console.error('Caught on try - onSubmit:', e);
+        setApp((prevState) => ({
+          ...prevState,
+          isLoading: false,
+          isError: true,
+          error: { status: 0, message: 'Unexpected error' },
+        }));
+      }
+    },
+    [search, setLocalState]
+  );
+
+  const value = useMemo(() => ({ app, detailedView }), [app, detailedView]);
+  const actions = useMemo(
+    () => ({
+      onChange,
+      onSubmit,
+      changePage,
+      toggleView,
+      fetchDetailedHandle,
+    }),
+    [onChange, onSubmit, toggleView, changePage, fetchDetailedHandle]
+  );
+
+  return (
+    <PokemonContext.Provider value={value}>
+      <ActionsContext.Provider value={actions}>
+        {children}
+      </ActionsContext.Provider>
+    </PokemonContext.Provider>
+  );
+};
```

Comment:
There is a lot of repeated logic for setting isLoading, handling errors, and updating app state after API calls (fetchDetailedHandle, changePage, both initializeData effects, and onSubmit). Consider extracting this into helper functions or a generic API handler to reduce duplication and make future changes easier to maintain.